File: wx-me/trunk/src/RetroMessenger.cpp
/****************************************************************************
* RetroMessenger - Serverless & Secure Instant Messenger
* build with wxWidgets gui toolkit based on libretroshare (c++)
*
* Copyright (c) 2008 Ryan Pusztai, Mahendra Panpalia and the RetroMessenger Team
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
* http://www.gnu.org/licenses/gpl-2.0.html
*
* Please report all praise, requests, bugs and problems to the project
* team & admins: http://sourceforge.net/projects/retromessenger
*
* For the libretroshare library see the updates here:
* http://sourceforge.net/projects/retroshare/
*
* You can find us listed at our projectpage. New team members are welcome.
* Please keep in touch, if you use this code for other projects.
*
* Website:
*	   http://sourceforge.net/projects/retromessenger/
*      http://retromessenger.sf.net
*****************************************************************************/

#include "RetroMessenger.h"
#include "RetroMessengerMainFrame.h"
#include "AppData.h"

// RetroShare
#include <rsiface/rsinit.h>

#include <openssl/ssl.h>

// wxWidgets
#include <wx/clipbrd.h>

IMPLEMENT_APP( MyApp );

bool MyApp::OnInit()
{
	// Initialize all image handlers.
	wxInitAllImageHandlers();
	
	// Initialize application data.
	m_globalData = std::auto_ptr< ApplicationDataBase >(AppData);

	// Load the saved settings.
	AppData->LoadFromFile();
	
	// Initialize OpenSSL libraries.
	SSL_load_error_strings();
	SSL_library_init();

	int rsArgc = wxTheApp->argc;
	std::vector< std::vector< char > > buffer;
	std::vector< char* > ptrs = wxCharPtrPtrToCharPtrPtr( wxTheApp->argv, rsArgc,  buffer );
	char** rsArgv = &ptrs[0];
	
	RsInit::SetBasedir(".");	

	// Initialize RsConfig
	RsInit::InitRsConfig();
	
	// Initialize RetroShare
	RsInit::InitRetroShare( rsArgc, rsArgv );
	
	// Make a main frame so we can add the friends.
	RetroMessengerMainFrame* frame = new RetroMessengerMainFrame();
	AppData->SetMainFrame( frame );
	SetTopWindow( frame );
	frame->Show();

	return true;
}

int MyApp::OnExit()
{
	AppData->SaveToFile();
	
	if( !wxTheClipboard->IsOpened() )
	{
        if ( !wxTheClipboard->Open() )
        {
            return wxApp::OnExit();
        }
	}

    // Allow clipboard data to persist after close
    wxTheClipboard->Flush();
    wxTheClipboard->Close();
	
	return wxApp::OnExit();
}

std::vector< char* > MyApp::wxCharPtrPtrToCharPtrPtr( wxChar** wxCharArray, int size, std::vector< std::vector< char > >& buffer )
{
	std::vector< char* > ptrs;

	// Clear the buffer.
	buffer.clear();

	for ( int i = 0; i < size; ++i )
	{
		// Create a wxString for the conversion to ASCII.
		wxString wxTmp( wxCharArray[i] );
		// Put the ASCII characters into a string to hold them long enough to push them into the buffer.
		std::string cTmp( wxTmp.ToAscii() );
		// Use a vector to make sure the memory will be contigious. Like the char** expects.
		std::vector< char > arg( cTmp.begin(), cTmp.end() );
		// Add the NULL char.
		arg.push_back( '\0' );
		// Push it into the vector passed into the function so that they are not destroyed after the function exits.
		buffer.push_back( arg );
		// Add it to the vector of pointers to the locations in the buffer vector.
		ptrs.push_back( &( buffer[i] )[0] );
	}

	return ptrs;
}

File: wx-me/trunk/retromessenger.lua
-- ---------------------------------------------------------------------------
-- retromessenger.lua - Premake script to build the GUI portion of
--
-- Copyright (c) 2008 Ryan Pusztai & RetroMessenger Team.
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
-- Please report all praise, requests, bugs and problems to the project
-- team & admins: http://sourceforge.net/projects/retromessenger
-- ---------------------------------------------------------------------------

-- GENERAL SETUP -------------------------------------------------------------
--
package.name								= "RetroMessenger"
package.language							= "c++"
package.kind								= "winexe"
package.target								= package.name
package.config["Debug"].target				= package.target .. "d"

-- PRE-BUILD SETUP -----------------------------------------------------------
--

-- HELPER FUNCTIONS -----------------------------------------------------------
--
function RemoveFiles( fileList )
	if type( fileList ) ~= "table" then
		error( "Type missmatch error occured in RemoveFiles()" )
	end

	for _, fileName in ipairs( fileList ) do
		if os.fileexists( fileName ) then
			os.remove( fileName )
		end
	end
end

-- COMPILER SETTINGS ----------------------------------------------------------
--
-- Build Flags
package.buildflags							= { "no-main" }
if options["dynamic-runtime"] then
	package.config["Release"].buildflags	= { "no-symbols", "optimize" }
else
	package.config["Debug"].buildflags		= { "static-runtime" }
	package.config["Release"].buildflags	= { "static-runtime", "no-symbols", "optimize" }
end

-- Defined Symbols
package.config["Debug"].defines				= { "DEBUG", "_DEBUG" }
package.config["Release"].defines			= { "NDEBUG" }

-- Files
package.files								= { matchfiles( "src/*.cpp", "src/*.h" ) }

package.excludes							= {}


-- LINKER SETTINGS 

if (windows) then

-- Include paths
package.includepaths 						= {
--												"libraries/kadc",
												"libraries/miniupnpc-1.3",
												"libraries/pthreads-w32-2-8-0-release",
--												"libraries/zlib-1.2.3",
												"libraries/openssl-0.9.8k/include",
												"libraries/libgpg-error-1.7/src",
												"libraries/gpgme-1.1.8/src",
												"libraries/retroshare"
											  }


package.config["Debug"].links =  {
								"libretroshared",
								"libgpg-error",
								"libgpgme",
								"libssl",
								"libcrypto",
								"libcrypt32-cygwin",
--								"kadc",
								"libminiupnpc",	
--								"libz",
								"libiphlpapi",
								"libpthreadGC2d"
								}
							
package.config["Release"].links =  {
								"libretroshare",
								"libgpg-error",
								"libgpgme",
								"libssl",
								"libcrypto",
								"libcrypt32-cygwin",
--								"kadc",
								"libminiupnpc",	
--								"libz",
								"libiphlpapi",
								"libpthreadGC2d"
							}


-- Linker directory paths.
package.libpaths							= {
												"libraries/retroshare/lib",
												"libraries/lib/windows"							
--												"libraries/kadc"
--												"libraries/lib/windows/libminiupnpc-1.3",
											  }
end

--------------

if (linux) then

-- Include paths
package.includepaths 						= {
--												"libraries/kadc",
												"libraries/miniupnpc-1.3",
--												"libraries/zlib-1.2.3",
												"libraries/openssl-0.9.8k/include",
												"libraries/libgpg-error-1.7/src",
												"libraries/gpgme-1.1.8/src",
												"libraries/retroshare"
											  }


package.config["Debug"].links =  {
									"libretroshared",
									"libgpg-error",
									"libgpgme",
									"libssl",
									"libcrypto",
									"libminiupnpc"
								}
			

package.config["Release"].links =  {
									"libretroshare",
									"libgpg-error",
									"libgpgme",
									"libssl",
									"libcrypto",
									"libminiupnpc"	
								}

			
-- Linker directory paths.
package.libpaths							= {
										"libraries/lib/linux",
										"libraries/lib/linux/gpg-related"
									  }

end


-- COMPILER SPECIFIC SETUP ----------------------------------------------------
--
if target == "gnu" or string.find( target or "", ".*-gcc" ) then
	table.insert( package.buildflags, "extra-warnings" )
	table.insert( package.buildoptions, { "-W", "-Wno-deprecated" } )
	table.insert( package.config["Release"].buildoptions, "-fno-strict-aliasing" )
	-- Set object output directory.
	if options["unicode"] then
		package.objdir						= ".obju"
	else
		package.objdir						= ".obj"
	end
end

if target == "vs2005" or target == "vs2008" then
	-- Visual C++ 2005/2008
	table.insert( package.buildflags, "seh-exceptions" )
	table.insert( package.defines, "_CRT_SECURE_NO_DEPRECATE" )
end

if target == "vs2003" then
	-- Enable re-mapping of Windows exceptions.
	package.buildoptions = { "/EHa" }
	table.insert( package.buildflags, { "no-exceptions" } )
end

-- OPERATING SYSTEM SPECIFIC SETTINGS -----------------------------------------
--
if windows then														-- WINDOWS
	table.insert( package.files, { "res/RetroMessenger.rc", "res/RetroMessenger.exe.manifest" } )
	table.insert( package.defines, { "_WIN32", "WIN32", "_WINDOWS","WINDOWS_SYS" } )
	table.insert( package.links, { "psapi", "ws2_32", "version" } )
elseif linux then													-- LINUX
	table.insert( package.links, { "pthread", "dl" } )
else																-- MACOSX
end

-- Copy images directory into bin (without svn related files) -----------------------------------------
--
for _,v in ipairs(matchrecursive("images/*")) do

 if (string.sub(v,-9) ~= ".svn-base") and (string.sub(v,-7) ~= "entries") and (string.sub(v,-11) ~= "all-wcprops") then
		s = string.sub(v,0,string.len(v)-string.len(path.getname(v))-1)
		t = ""

		for w in string.gmatch(s, "(%w+)") do
			t = t.."/"..w
		
			if (os.direxists(project.bindir..t) == false) then
				os.mkdir(project.bindir..t)
			end
		end

	os.copyfile(v,project.bindir.."/"..v)
	end
end


-- WXWIDGETS PACKAGE SETUP ----------------------------------------------------
--
-- ConfigureWxWidgets( package, altTargetName, wxVer, wxVerMinor, useAltTargetNameOnly )
ConfigureWxWidgets( package, package.name, "28", "8" )
